lib/lookup/src/lookup_v2/mod.rs

```diff
@@ -186,6 +200,33 @@ impl<'a, 'b: 'a> From<&'b OwnedSegment> for BorrowedSegment<'a> {
     }
 }
 
+impl Display for OwnedSegment {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            OwnedSegment::Field(field) => {
+                static UNESCAPED_PATTERN: Lazy<Regex> =
+                    Lazy::new(|| Regex::new(r"^[A-Za-z_0-9@]*$").unwrap());
```

Comment:
This seems somewhat brittle.

I would expect the `OwnedSegment` (or `Field`?) types to encode whether the segment needs to be escaped. Either through enum variants, or a flag in a struct somewhere.

Reply:
If this is implemented by adding the data do `OwnedSegment` or `Field` and then looking it up, just be _very_ careful about how it's implemented. The parser / iterating over the segments is a _very_ hot path and performance is critical. The `Display` path is hopefully not performance sensitive, so I would prefer to defer calculations to it when possible.

Reply:
Implemented in 13a88eeefc6242a38fd22cf9ec839b06f586776b. It figures out if the quote is needed and bubbles that information up. Adding that state should not have any performance impact as far as I'm aware

Reply:
(Since it's doing equality checks on the character for `"` or `\` in match branches before, that might now even have any performance impact, depending on the code that the compiler generates for this switch.)

Reply:
there are some benches in `lib/lookup/benches/lookup.rs` if you want to run that and compare to before this change. Your changes seem reasonable, but I'd prefer to double check since this is used so often.

Reply:
Looks good to me @pablosichert 👍 

One suggestion, similar to what we did in VRL, you might want to lower the "requires quoting" variant into a `Field` enum.

This means you'd keep `Field`, `Index`, and `Invalid` for `OwnedSegment`, but `Field` now takes a `Field` type instead of `String`, which is an enum over `Quoted` and `Regular` or something similar.

e.g.:

```rust
enum OwnedSegment {
	Field(Field)
	...
}

enum Field {
	Quoted(String),
	Regular(String),
}
```

Reply:
I really don't like this change. Quoting is purely a visual thing, and adding it into the data of a "field" is unnecessary. It will make specifying paths in the future more difficult too. For example, if in code you want to specify to place something in metadata, (using a macro I've implemented in a current PR), it would look like `path!("metadata", "foo")`. This should generate a static array of `BorrowedSegment`, which is trivial before this change. With this change, we would either need a function to determine if any of those fields need to be quoted, or the user need to supply it which is really error prone. If we need a function to determine if something is quoted, then we can just use that in the `Display` implementation and not complicate the `Field` values.
Similarly, the `BorrowedSegment::field()` function is not assuming the field is _not_ quoted, and there's no alternative for a quoted alternative.

Reply:
I had the pain of going through all the issues that @fuchsnj mentioned in my approach in d5655ef3f896f2f0799ae79ff17ddd684c011581...

Ideally I'd really like to have the escape information without any additional processing from within `Serialize`, but both the runtime cost and carefullness that instantiating `Segments` all over codebase would require really doesn't seem worth the cost

Reply:
This doesn't really change the concerns about maintainability, but at least this is not using a regex any more: 1bfbe232ac6cde4d7f01a270bc3473ab0e11a4b3